File: hmh-frontend/src/api/companies.ts
import client from "./client";
import { Supplier } from "./suppliers";
import { Company, CompanyCreate } from "./projects";

export type { Company, CompanyCreate };

export interface CompanyWithSuppliers extends Company {
  suppliers: Supplier[];
}

export interface CompanyUpdate {
  name?: string;
  registration_number?: string | null;
  contact_email?: string | null;
  contact_phone?: string | null;
  address?: string | null;
  notes?: string | null;
}

export const companiesApi = {
  list: async (): Promise<Company[]> => {
    const res = await client.get<{ data: Company[] }>("/companies/");
    return res.data.data;
  },

  get: async (id: string): Promise<CompanyWithSuppliers> => {
    const res = await client.get<{ data: CompanyWithSuppliers }>(`/companies/${id}`);
    return res.data.data;
  },

  create: async (body: CompanyCreate): Promise<Company> => {
    const res = await client.post<{ data: Company }>("/companies/", body);
    return res.data.data;
  },

  update: async (id: string, body: CompanyUpdate): Promise<Company> => {
    const res = await client.patch<{ data: Company }>(`/companies/${id}`, body);
    return res.data.data;
  },

  delete: async (id: string): Promise<void> => {
    await client.delete(`/companies/${id}`);
  },

  linkSupplier: async (companyId: string, supplierId: string): Promise<void> => {
    await client.post(`/companies/${companyId}/suppliers/${supplierId}`);
  },

  unlinkSupplier: async (companyId: string, supplierId: string): Promise<void> => {
    await client.delete(`/companies/${companyId}/suppliers/${supplierId}`);
  },

  getProjectSuppliers: async (projectId: string): Promise<Supplier[]> => {
    const res = await client.get<{ data: Supplier[] }>(`/companies/project/${projectId}/suppliers`);
    return res.data.data;
  },
};

File: hmh-backend/app/api/v1/companies.py
"""Company CRUD + supplier-link routes."""

import uuid

from fastapi import APIRouter

from app.dependencies import ALL_ROLES, DbSession, OFFICE_ADMIN_AND_ABOVE
from app.schemas.common import ApiSuccess
from app.schemas.company import (
    CompanyCreate,
    CompanyRead,
    CompanySupplierLinkRead,
    CompanyUpdate,
    CompanyWithSuppliersRead,
)
from app.schemas.supplier import SupplierRead
from app.services import company_service

router = APIRouter(prefix="/companies", tags=["companies"])


@router.get("/", response_model=ApiSuccess[list[CompanyRead]], dependencies=[ALL_ROLES])
def list_companies(db: DbSession):
    companies = company_service.list_companies(db)
    return ApiSuccess(data=[CompanyRead.model_validate(c) for c in companies])


@router.post(
    "/",
    response_model=ApiSuccess[CompanyRead],
    status_code=201,
    dependencies=[OFFICE_ADMIN_AND_ABOVE],
)
def create_company(body: CompanyCreate, db: DbSession):
    company = company_service.create_company(db, body)
    return ApiSuccess(data=CompanyRead.model_validate(company), message="Company created.")


@router.get("/{company_id}", response_model=ApiSuccess[CompanyWithSuppliersRead], dependencies=[ALL_ROLES])
def get_company(company_id: uuid.UUID, db: DbSession):
    company = company_service.get_company(db, company_id)
    suppliers = company_service.get_company_suppliers(db, company_id)
    result = CompanyWithSuppliersRead(
        **CompanyRead.model_validate(company).model_dump(),
        suppliers=[SupplierRead.model_validate(s) for s in suppliers],
    )
    return ApiSuccess(data=result)


@router.patch(
    "/{company_id}",
    response_model=ApiSuccess[CompanyRead],
    dependencies=[OFFICE_ADMIN_AND_ABOVE],
)
def update_company(company_id: uuid.UUID, body: CompanyUpdate, db: DbSession):
    company = company_service.update_company(db, company_id, body)
    return ApiSuccess(data=CompanyRead.model_validate(company), message="Company updated.")


@router.delete(
    "/{company_id}",
    response_model=ApiSuccess[dict],
    dependencies=[OFFICE_ADMIN_AND_ABOVE],
)
def delete_company(company_id: uuid.UUID, db: DbSession):
    result = company_service.delete_company(db, company_id)
    return ApiSuccess(data=result, message=f"Company '{result['deleted_company']}' deleted.")


@router.post(
    "/{company_id}/suppliers/{supplier_id}",
    response_model=ApiSuccess[CompanySupplierLinkRead],
    status_code=201,
    dependencies=[OFFICE_ADMIN_AND_ABOVE],
)
def link_supplier(company_id: uuid.UUID, supplier_id: uuid.UUID, db: DbSession):
    link = company_service.link_supplier(db, company_id, supplier_id)
    return ApiSuccess(
        data=CompanySupplierLinkRead.model_validate(link),
        message="Supplier linked to company.",
    )


@router.delete(
    "/{company_id}/suppliers/{supplier_id}",
    response_model=ApiSuccess[dict],
    dependencies=[OFFICE_ADMIN_AND_ABOVE],
)
def unlink_supplier(company_id: uuid.UUID, supplier_id: uuid.UUID, db: DbSession):
    company_service.unlink_supplier(db, company_id, supplier_id)
    return ApiSuccess(data={}, message="Supplier unlinked.")


@router.get(
    "/project/{project_id}/suppliers",
    response_model=ApiSuccess[list[SupplierRead]],
    dependencies=[ALL_ROLES],
)
def get_project_suppliers(project_id: uuid.UUID, db: DbSession):
    """
    Returns the allowed suppliers for a project.
    If the project has a linked company, returns only that company's suppliers.
    If no company is linked, returns all active suppliers.
    """
    from app.models.supplier import Supplier as SupplierModel

    suppliers = company_service.get_project_suppliers(db, project_id)
    if suppliers is None:
        # No company restriction — return all active suppliers
        suppliers = (
            db.query(SupplierModel)
            .filter(SupplierModel.is_active == True)  # noqa: E712
            .order_by(SupplierModel.name)
            .all()
        )
    return ApiSuccess(data=[SupplierRead.model_validate(s) for s in suppliers])
